pypesto/objective/amici.py

```diff
@@ -423,10 +432,27 @@ def custom_timepoints(
 
         amici_objective = copy.deepcopy(self)
 
+        if timepoints is not None:
+            if len(timepoints) != len(amici_objective.edatas):
+                raise ValueError(
+                    'The number of condition-specific timepoints `timepoints` '
+                    'does not match the number of experimental conditions.\n'
+                    f'Number of provided timepoints: {len(timepoints)}. '
+                    'Number of experimental conditions: '
+                    f'{len(amici_objective.edatas)}.'
+                )
+            custom_timepoints = timepoints
+        else:
+            custom_timepoints = [
+                copy.deepcopy(timepoints_global)
+                for _ in range(len(amici_objective.edatas))
+            ]
+
+        amici_objective.custom_timepoints = custom_timepoints
+
         for index in range(len(amici_objective.edatas)):
-            if timepoints_global is not None:
-                amici_objective.edatas[index].setTimepoints(timepoints_global)
-            else:
-                amici_objective.edatas[index].setTimepoints(timepoints[index])
+            amici_objective.edatas[index].setTimepoints(
```

Comment:
I think this should be implemented as separate method (`apply_custom_timepoints`) that can also be called from `__setstate__`